File: ensark/src/main/java/com/elitetech_inc/ensarkbank/customer_management/customer/service/CustomerService.java
package com.elitetech_inc.ensarkbank.customer_management.customer.service;

import com.elitetech_inc.ensarkbank.common.enums.DocumentType;
import com.elitetech_inc.ensarkbank.customer_management.customer.dto.request.CustomerRequest;
import com.elitetech_inc.ensarkbank.customer_management.customer.dto.response.CustomerResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;

public interface CustomerService {

    CustomerResponse create(CustomerRequest request, MultipartFile profile, Map<DocumentType, MultipartFile> documents);

    CustomerResponse findById(Long id);

    List<CustomerResponse> getAll();

    List<CustomerResponse> getAllByBranchIds(List<Long> branchIds);

    boolean customerEmailExists(String email);

    List<CustomerResponse> search(String query);

    CustomerResponse updateByEmployee(Long id, CustomerRequest request, MultipartFile profile);

    CustomerResponse updatePassword(Long id, String oldPassword, String newPassword);

    CustomerResponse updateProfilePicture(Long id, MultipartFile profile);

    CustomerResponse findByEmail(String email);

    void delete(Long id);

    void deactivate(Long id);

    CustomerResponse findByPhone(String phone);

    boolean phoneExistsCheck(String phone);

    List<CustomerResponse> findByStatus(com.elitetech_inc.ensarkbank.common.enums.CustomerStatus status);
}

File: ensark-frontend/src/app/core/services/customer.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CustomerRequest, CustomerResponse } from '../models/customer.models';
import { CustomerStatus } from '../models/enums';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CustomerService {
  private http = inject(HttpClient);
  private apiUrl = environment.apiUrl + 'customer/';

  getAll(): Observable<CustomerResponse[]> {
    return this.http.get<CustomerResponse[]>(`${this.apiUrl}`);
  }

  getById(id: number): Observable<CustomerResponse> {
    return this.http.get<CustomerResponse>(`${this.apiUrl}${id}`);
  }

  getByEmail(email: string): Observable<CustomerResponse> {
    return this.http.get<CustomerResponse>(`${this.apiUrl}email/${email}`);
  }

  search(query: string): Observable<CustomerResponse[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<CustomerResponse[]>(`${this.apiUrl}search`, { params });
  }

  emailExists(email: string): Observable<boolean> {
    const params = new HttpParams().set('email', email);
    return this.http.get<boolean>(`${this.apiUrl}exists`, { params });
  }

  update(id: number, data: CustomerRequest, profile?: File): Observable<CustomerResponse> {
    const formData = new FormData();
    formData.append('data', JSON.stringify(data));
    if (profile) {
      // CustomerController.updateByEmployee expects @RequestPart("profile") — the previous
      // 'profilePicture' key silently dropped every profile-photo update since the part was
      // optional server-side and therefore never rejected, just ignored.
      formData.append('profile', profile);
    }
    return this.http.put<CustomerResponse>(`${this.apiUrl}${id}`, formData);
  }

  updateProfilePicture(id: number, profile: File): Observable<CustomerResponse> {
    const formData = new FormData();
    formData.append('profile', profile);
    return this.http.put<CustomerResponse>(`${this.apiUrl}${id}/profile`, formData);
  }

  updatePassword(id: number, oldPass: string, newPass: string): Observable<void> {
    const params = new HttpParams()
      .set('oldPass', oldPass)
      .set('newPass', newPass);
    return this.http.patch<void>(`${this.apiUrl}${id}/password`, null, { params });
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}${id}`);
  }

  deactivate(id: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}${id}/deactivate`, {});
  }

  findByPhone(phone: string): Observable<CustomerResponse> {
    return this.http.get<CustomerResponse>(`${this.apiUrl}phone/${phone}`);
  }

  phoneExists(phone: string): Observable<boolean> {
    const params = new HttpParams().set('phone', phone);
    return this.http.get<boolean>(`${this.apiUrl}phone-exists`, { params });
  }

  findByStatus(status: CustomerStatus): Observable<CustomerResponse[]> {
    return this.http.get<CustomerResponse[]>(`${this.apiUrl}status/${status}`);
  }
}
